feat: series and book files deletion

Snd-R opened the pull request:
Allows users to delete series and book files 

Closes: #731

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/rest/BookController.kt:
```diff
@@ -550,6 +551,25 @@ class BookController(
+        throw ResponseStatusException(HttpStatus.CONFLICT)
+      }
+
+      bookLifecycle.deleteOne(book)
+      taskReceiver.deleteFile(book.url)
```
The first operation is synchronous, while the second is asynchronous, which could lead to consistency issues.

There are also side effects when deleting a book, which are handled by the scan, for instance the reordering of the remaining books in a series, or the series metadata aggregation.

We should also consider deleting the book's sidecar files.

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/rest/BookController.kt:
```diff
@@ -550,6 +551,25 @@ class BookController(
+  ) {
+    bookRepository.findByIdOrNull(bookId)?.let { book ->
+      if (!principal.user.canAccessBook(book)) throw ResponseStatusException(HttpStatus.FORBIDDEN)
+      if (!Files.isWritable(book.path)) {
+        logger.warn { "Cannot delete ${book.path}: Permission denied" }
+        throw ResponseStatusException(HttpStatus.CONFLICT)
```
I would say that this type of control lies in the domain service layer. From the API standpoint, if you return NO CONTENT, then it's a command, and the command is valid if you have the rights (ie ADMIN).

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/rest/SeriesController.kt:
```diff
@@ -682,4 +684,25 @@ class SeriesController(
+      }
+
+      seriesLifecycle.deleteOne(series)
+      books.forEach { taskReceiver.deleteFile(it.url) }
```
Deleting only the files contained in the folder would leave the folder itself, and the sidecars.

gotson reviewed (commented):
Thanks for the PR!

I am not sure about the ordering of actions though.

In my opinion, we should:
- have the deletion logic in the domain services
- when deleting a book, we should either delete the file then trigger a scan, or have some special handling, like in `BookImporter.kt` when upgrading a book.
- when deleting a series, we should delete all the books within, then check whether the folder is empty (it could contain other series), and if yes, delete it too.

The backend part would also need to be covered by tests.

Snd-R commented:
Thanks for the feedback! I refactored the code with the following changes:

- Made the whole process asynchronous like it done with book imports.
- Moved deletion logic into services
- When files are deleted then Library rescan will be triggered

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/SeriesController.kt:
```diff
@@ -682,4 +683,20 @@ class SeriesController(
+        seriesId = seriesId,
+        priority = HIGHEST_PRIORITY,
+      )
+    } catch (e: Exception) {
```
the method called doesn't throw anything, there's no need for try/catch

gotson commented on komga/src/main/kotlin/org/gotson/komga/application/tasks/TaskHandler.kt:
```diff
@@ -121,6 +123,20 @@ class TaskHandler(
+          is Task.DeleteBook -> {
+            bookRepository.findByIdOrNull(task.bookId)?.let { book ->
+              bookLifecycle.deleteBookFiles(book)
+              taskReceiver.scanLibrary(book.libraryId)
```
we should trigger the scan task with the same priority, so it's executed straight away. If not, and there's plenty of tasks in queue already, the scan could get delayed a lot, which would provide a bad UX IMHO.

gotson commented on komga/src/main/kotlin/org/gotson/komga/application/tasks/TaskHandler.kt:
```diff
@@ -121,6 +123,20 @@ class TaskHandler(
+          is Task.DeleteSeries -> {
+            seriesRepository.findByIdOrNull(task.seriesId)?.let { series ->
+              seriesLifecycle.deleteSeriesFiles(series)
+              taskReceiver.scanLibrary(series.libraryId)
```
same as above

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/service/BookLifecycle.kt:
```diff
@@ -272,4 +281,23 @@ class BookLifecycle(
+      throw FileNotFoundException("File is not accessible : ${book.path}").withCode("ERR_1018")
+
+    val thumbnails = thumbnailBookRepository.findAllByBookId(book.id)
+      .filter { it.type == ThumbnailBook.Type.SIDECAR }
```
it's best to implement the filtering in the repository method, especially since there is already `findAllByBookIdAndType`

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/service/BookLifecycle.kt:
```diff
@@ -272,4 +281,23 @@ class BookLifecycle(
+      .filter { it.type == ThumbnailBook.Type.SIDECAR }
+      .mapNotNull { it.url?.toURI()?.toPath() }
+
+    thumbnails.find { it.notExists() || !it.isWritable() }?.let {
+      throw FileNotFoundException("File is not accessible : ${book.path}").withCode("ERR_1018")
+    }
```
that means the book deletion fails if you can't find one of the sidecar ? That's drastic, sidecars are updated only during scans, so we don't really care if they are there or not. If they are, we delete them, if not, the Thumbnail entry will get cleaned up eventually.

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/service/SeriesLifecycle.kt:
```diff
@@ -278,6 +287,27 @@ class SeriesLifecycle(
+      .filter { it.type == ThumbnailSeries.Type.SIDECAR }
+      .map { it.url!!.toURI().toPath() }
+
+    thumbnails.find { it.notExists() || !it.isWritable() }?.let {
+      throw FileNotFoundException("File is not accessible : $it").withCode("ERR_1018")
+    }
```
same as for books

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/persistence/ReadProgressRepository.kt:
```diff
@@ -18,6 +18,7 @@ interface ReadProgressRepository {
   fun deleteByBookId(bookId: String)
   fun deleteByBookIds(bookIds: Collection<String>)
   fun deleteByBookIdsAndUserId(bookIds: Collection<String>, userId: String)
+  fun deleteBySeriesId(seriesId: String)
```
we may not need that anymore, do we?

gotson commented on komga/src/main/kotlin/org/gotson/komga/infrastructure/jooq/ReadProgressDao.kt:
```diff
@@ -118,6 +118,10 @@ class ReadProgressDao(
     aggregateSeriesProgress(bookIds)
   }
 
+  override fun deleteBySeriesId(seriesId: String) {
```
we may not need that anymore, do we?

gotson commented on komga/src/test/kotlin/org/gotson/komga/domain/service/BookLifecycleTest.kt:
```diff
@@ -130,4 +139,136 @@ class BookLifecycleTest(
+  }
+
+  @Test
+  fun `given a book and a non-existent sidecar file when deleting book then exception is thrown and no files are deleted`() {
```
as mentioned in the service class, i think this is not a good behaviour, book should be deleted anyhow

gotson commented on komga/src/main/kotlin/org/gotson/komga/application/tasks/TaskReceiver.kt:
```diff
@@ -42,7 +42,7 @@ class TaskReceiver(
 
-  fun scanLibrary(libraryId: String) {
+  fun scanLibrary(libraryId: String, priority: Int = DEFAULT_PRIORITY) {
     submitTask(Task.ScanLibrary(libraryId))
```
```suggestion
submitTask(Task.ScanLibrary(libraryId, priority))
```

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/service/BookLifecycle.kt:
```diff
@@ -272,4 +282,19 @@ class BookLifecycle(
+      .filter { it.exists() && it.isWritable() }
+
+    book.path.deleteIfExists()
+    thumbnails.forEach(Path::deleteIfExists)
```
```suggestion
thumbnails.forEach { it.deleteIfExists() }
```

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/service/SeriesLifecycle.kt:
```diff
@@ -278,6 +287,22 @@ class SeriesLifecycle(
+      throw FileNotFoundException("File is not accessible : ${series.path}").withCode("ERR_1018")
+
+    val thumbnails = thumbnailsSeriesRepository.findAllBySeriesIdIdAndType(series.id, ThumbnailSeries.Type.SIDECAR)
+      .map { it.url!!.toURI().toPath() }
```
```suggestion
.mapNotNull { it.url?.toURI()?.toPath() }
```

gotson commented on komga/src/main/kotlin/org/gotson/komga/domain/service/SeriesLifecycle.kt:
```diff
@@ -278,6 +287,22 @@ class SeriesLifecycle(
+      .map { it.url!!.toURI().toPath() }
+      .filter { it.exists() && it.isWritable() }
+
+    val books = bookRepository.findAllBySeriesId(series.id)
+    books.forEach { bookLifecycle.deleteBookFiles(it) }
```
```suggestion
bookRepository.findAllBySeriesId(series.id)
      .forEach { bookLifecycle.deleteBookFiles(it) }
```

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/BookController.kt:
```diff
@@ -543,6 +543,19 @@ class BookController(
     }
   }
 
+  @DeleteMapping("api/v1/books/{bookId}")
```
I am conflicted about that one. We are not deleting the Book entity, but the underlying file.

I think it would make more sense to use `api/v1/books/{bookId}/file` as the path instead.

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/BookController.kt:
```diff
@@ -543,6 +543,19 @@ class BookController(
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  fun deleteBook(
+    @PathVariable bookId: String,
+    @AuthenticationPrincipal principal: KomgaPrincipal
```
this is unused

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/SeriesController.kt:
```diff
@@ -682,4 +683,16 @@ class SeriesController(
+
+  @DeleteMapping("v1/series/{seriesId}")
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  fun deleteSeries(
```
You are missing the authorization for `ROLE_ADMIN`

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/SeriesController.kt:
```diff
@@ -682,4 +683,16 @@ class SeriesController(
       .body(streamingResponse)
   }
+
+  @DeleteMapping("v1/series/{seriesId}")
```
Same concern as for books, we are not deleting the entity, but the underlying files, would make more sense to use the `/file` endpoint for deletion.

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/SeriesController.kt:
```diff
@@ -682,4 +683,16 @@ class SeriesController(
+  @ResponseStatus(HttpStatus.ACCEPTED)
+  fun deleteSeries(
+    @PathVariable seriesId: String,
+    @AuthenticationPrincipal principal: KomgaPrincipal
```
this is unused

gotson commented on komga-webui/src/locales/en.json:
```diff
@@ -272,6 +272,15 @@
     },
+    "delete_book": {
+      "button_confirm": "Delete",
+      "confirm_delete": "Yes, delete book \"{name}\" and it's files",
```
```suggestion
"confirm_delete": "Yes, delete book \"{name}\" and its files",
```

gotson commented on komga-webui/src/locales/en.json:
```diff
@@ -302,6 +311,14 @@
     },
+    "delete_series": {
+      "button_confirm": "Delete",
+      "confirm_delete": "Yes, delete series \"{name}\" and it's files",
```
```suggestion
"confirm_delete": "Yes, delete series \"{name}\" and its files",
```

gotson reviewed (changes requested):
I suggested a couple of changes, but overall this PR is of very goof quality!

gotson commented on komga/src/main/kotlin/org/gotson/komga/interfaces/api/rest/BookController.kt:
```diff
@@ -543,12 +543,11 @@ class BookController(
   }
 
-  @DeleteMapping("api/v1/books/{bookId}")
+  @DeleteMapping("api/v1/books/{bookId}/file")
```
I suppose you will need to update the path in the front-end too 😅

gotson commented:
Thansk a lot, I'll review again tomorrow morning but that should be good to go.
